doc/source/serve/deployment-graph.md

```diff
@@ -0,0 +1,565 @@
+```
+
+- Building a graph:
+    - A deployment node is created from `@serve.deployment` decorated function or class. 
+    - You can construct different deployment nodes from same class or function.
+    - Deployment nodes can be used as input args in .bind() of other nodes in the DAG. 
+    - Multiple nodes in the deployment graph naturally forms a DAG structure.
+- Accessing input:
+    - Same input or output can be used in multiple nodes in the DAG.
+    - Deployment nodes can access partial user input.
+- Dynamic control flow:
+    - A deployment node can call into other nodes in the deployment graph.
+    - You can use the dynamic calling to perform control flow operation that's hard to expressive in traditional DAG.
+- Running a DAG:
+    - A node can be called either sync or async.
+
++++
+
+![deployment graph](https://github.com/ray-project/images/blob/master/docs/serve/deployment_graph.png?raw=true)
+
++++
+
+### Step 1: User InputNode and preprocessor 
+
+Let's start with the first layer of DAG: Building user input to two preprocessor functions, where each function receives parts of the input data. For simplicity, we use the same existing `@serve.deployment` decorator on an async function body.
+
++++
+
+```python
+from ray import serve
+# We will later move Ray DAG related components out of experimental in later stable release
+from ray.experimental.dag.input_node import InputNode
+
+@serve.deployment
+async def preprocessor(input_data: str):
+    """Simple feature processing that converts str to int"""
+    time.sleep(0.1) # Manual delay for blocking computation
+    return int(input_data)
+
+@serve.deployment
+async def avg_preprocessor(input_data):
+    """Simple feature processing that returns average of input list as float."""
+    time.sleep(0.15) # Manual delay for blocking computation
+    return sum(input_data) / len(input_data)
+
+# DAG building
+with InputNode() as dag_input:
+    # Partial access of user input by index
+    preprocessed_1 = preprocessor.bind(dag_input[0])  
+    preprocessed_2 = avg_preprocessor.bind(dag_input[1])
+```
+
++++
+
+There are two new APIs used in the DAG building stage: `InputNode()` and `bind()`.
+
+### **```InputNode()```** : User input of the graph
+
+```InputNode``` is a special singleton in DAG building that's only relevant to it's runtime call behavior. Even though all decorated classes or functions can be reused in arbitrary way to facilitate DAG building where the root DAGNode forms the graph with its children, in each deployment graph there should be one and only one InputNode used.
+
+```InputNode``` value is fulfilled and replaced by user input at runtime, therefore it takes no argument when being constructed.
+
+```InputNode``` should only be created within a context manager to indicate there's one and only one `InputNode` for each graph, and will throw exception if directly created, or multiple is found.
```

Comment:
i can do the shortening, but i don't think `InputNode` is the root node, i think it's the opposite leaf node that always gets resolved first when we try to execute the DAG. From user perspective, the root node is the entrypoint node of user choice, such as `combine.run()`; For execution, `InputNode` gets replaced by user value in first layer. So for user facing docs I think we probably want to let user think InputNode is just your inputs in arg rather than root of the  dag.